### PR TITLE
Add Schema to SQL Server and PostgreSQL dialect

### LINKED ISSUE #379: Violation of PRIMARY KEY constraint 'PK__t4h_Docu__3214EC07B34ED974'. Cannot insert duplicate key in object 'dbo.t4h_Document'. The duplicate key value is (81)
This Issue was first reported on Orchard Core:
https://github.com/OrchardCMS/OrchardCore/issues/9088

### In short:
Trying to create new Users in Orchard Core did not work and created the error and Stack-Trace reported to the ticket above.
### Context:
1. Two Instances of Orchard Core were running at the same time on the same database, one of them in Visual Studio, one as a live preview.
2. `Identifiers` table seemed to have a value above `max id` of documents, but the ids causing this error were in the range below `max id`
3. After stopping both instances and just starting one, the system recovered after one insuccessful try.
4. Even after this one instance recovered, the inserted new ids were below `max id` and the `nextval` value in `identifiers` pointed to the same value as when the system did not work.
5. Obviously, as lower ids could be inserted, there are unused ids below `max id` in my `documents` table
### Hypothesis
1. There is a mechanism (intended or not), that recovers unused ids below `max id`, but the ids that one instances claims in the range below `max id` are not represented in the database and therefore not communicated to other instances.

### Version
In my Orchard Core Version 1.0.0.0 RC2 stable, the used YesSql still seems to be way below the current version.

### PR DESCRIPTION
**Pull request by Skrypt**
Fixes #379

A Schema is a logical collection of database objects for these 2.

## TODO

- [x] Create Schema on the database from SchemaBuilder.
- [x] Make unit tests pass with a custom schema name in PostgreSQL
- [x] Make unit tests pass with a custom schema name in SQL Server
- [x] Make sure everything is backward compatible

### CODE REVIEW

File: src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs [hunk 1]
```diff
@@ -113,6 +113,7 @@ public PostgreSqlDialect()
         public override string RandomOrderByClause => "random()";
         public override bool SupportsIfExistsBeforeTableName => true;
         public override bool PrefixIndex => true;
+        public string DefaultSchema => "public";
```
Comment (sebastienros) [inline src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs:116]:
```suggestion
        public string Schema { get; set; }
        public const string DefaultSchema = "public";
```

File: src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs [hunk 2]
```diff
@@ -220,7 +221,7 @@ public override string QuoteForColumnName(string columnName)
 
         public override string QuoteForTableName(string tableName)
         {
-            return QuoteString + tableName + QuoteString;
+            return QuoteString + DefaultSchema + "." + tableName + QuoteString;
```
Comment (sebastienros) [inline src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs:224]:
```suggestion
            return QuoteString + Schema ?? DefaultSchema + QuoteString + "." + QuoteString + tableName + QuoteString;
```

File: src/YesSql.Provider.SqlServer/SqlServerDialect.cs [hunk 3]
```diff
@@ -109,6 +109,8 @@ public SqlServerDialect()
 
         public override byte DefaultDecimalScale => 5;
 
+        public string DefaultSchema => "dbo";
```
Comment (sebastienros) [inline src/YesSql.Provider.SqlServer/SqlServerDialect.cs:112]:
```suggestion
        public string Schema { get; set; }
        public const string DefaultSchema = "dbo";
```

File: src/YesSql.Provider.SqlServer/SqlServerDialect.cs [hunk 4]
```diff
@@ -208,7 +210,7 @@ public override string QuoteForColumnName(string columnName)
 
         public override string QuoteForTableName(string tableName)
         {
-            return "[" + tableName + "]";
+            return "[" + DefaultSchema + "." + tableName + "]";
```
Comment (sebastienros) [inline src/YesSql.Provider.SqlServer/SqlServerDialect.cs:213]:
```suggestion
            return "[" + Schema ?? DefaultSchema + "].[" + tableName + "]";
```

**Comment by Skrypt** [comment]
Will require some more thinking. There are some issues here and there because for example we can't do :  

`SELECT * FROM dbo.Table AS dbo.Table2`

So, I'm changing _dialect.QuoteForTableName(value) for _dialect.QuoteForColumnName(value) in some places so that it works but it might break for other types of databases.

image

**Comment by Skrypt** [comment]
image

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 5]
```diff
@@ -48,12 +48,12 @@ public SqlBuilder(string tablePrefix, ISqlDialect dialect)
         public void Table(string table, string alias = null)
         {
             FromSegments.Clear();
-            FromSegments.Add(_dialect.QuoteForTableName(_tablePrefix + table));
+            FromSegments.Add(_dialect.QuoteForColumnName(_tablePrefix + table));
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:51]: This one looks wrong

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 6]
```diff
@@ -93,14 +93,14 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
             }
             
             JoinSegments.Add(" INNER JOIN ");
-            JoinSegments.Add(_dialect.QuoteForTableName(_tablePrefix + table));
+            JoinSegments.Add(_dialect.QuoteForColumnName(_tablePrefix + table));
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:96]: Looks wrong

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 7]
```diff
@@ -93,14 +93,14 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
...
             if (!String.IsNullOrEmpty(alias))
             {
-                JoinSegments.AddRange(new[] { " AS ", _dialect.QuoteForTableName(alias) });
+                JoinSegments.AddRange(new[] { " AS ", _dialect.QuoteForColumnName(alias) });
             }
             JoinSegments.AddRange(new[] {
-                " ON ", _dialect.QuoteForTableName(onTable), ".", _dialect.QuoteForColumnName(onColumn),
-                " = ", _dialect.QuoteForTableName(toTable), ".", _dialect.QuoteForColumnName(toColumn)
+                " ON ", _dialect.QuoteForColumnName(onTable), ".", _dialect.QuoteForColumnName(onColumn),
+                " = ", _dialect.QuoteForColumnName(toTable), ".", _dialect.QuoteForColumnName(toColumn)
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:103]: Looks wrong

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 8]
```diff
@@ -160,7 +160,7 @@ public virtual string FormatColumn(string table, string column, bool isAlias = f
                 table = _tablePrefix + table;
             }
 
-            return _dialect.QuoteForTableName(table) + "." + column;
+            return _dialect.QuoteForColumnName(table) + "." + column;
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:163]: Looks wrong

**Comment by Skrypt** [comment]
Yes I know but see, it works still. So, while it is wrong it passes. But I need to find a better way, I agree.

**Comment by sebastienros** [comment]
It works because it just ignores the schema in this case, like it was doing before the PR, but that wrong, the goal is to specify the schema any time the table name is used as a table name.

**Comment by sebastienros** [comment]
I suggest to use docker to install Postres locally, I believe I added the command line to run in the unit test for that.

**Comment by sebastienros** [comment]
Ideally the unit tests dialects should be initialized with a different schema, some tests would probably fail.

File: src/YesSql.Core/Sql/SqlBuilder.cs (hunk 5 above)
Comment (Skrypt) [reply src/YesSql.Core/Sql/SqlBuilder.cs:51]: Here the issue is that it passes there when we create a table alias.
So it does something like : 

`SELECT * FROM [dbo].[Table] AS [dbo].[Table1] INNER JOIN [dbo].[JoinedTable] AS [dbo].[JoinedTable1] ON [dbo].[JoinedTable1].[Column] = [dbo].[Table1].[Column]`

This is invalid SQL to use the schema prefix on table aliases.

**Comment by Skrypt** [comment]
I just tested this on a local PostgreSQL docker instance. See the CI passes too.

**Comment by sebastienros** [comment]
The AS statements don't need the schema, they can use column quoting.

**Comment by Skrypt** [comment]
I did not analyze everything yet. Just a first pass to see if that would actually work. And it seems like it would. So that's at least a good sign.

File: src/YesSql.Provider.SqlServer/SqlServerDialect.cs [hunk 9]
```diff
@@ -206,9 +209,17 @@ public override string QuoteForColumnName(string columnName)
             return "[" + columnName + "]";
         }
 
-        public override string QuoteForTableName(string tableName)
+        public override string QuoteForTableName(string tableName, bool IncludeSchema = false)
         {
-            return "[" + tableName + "]";
+            var result = "[" + tableName + "]";
```
Comment (sebastienros) [inline src/YesSql.Provider.SqlServer/SqlServerDialect.cs:214]: How is it different from `QuoteForColumn` when `includeSchema` is false? I even question semantically. Meaning that it could be the same but we'd want to keep them with different methods, but here I don't think it's "a table without the schema" we want to render.
I will double check with your previous examples.

Comment (Skrypt) [reply src/YesSql.Provider.SqlServer/SqlServerDialect.cs:214]: Yes, that's why I had used QuoteForColumn because there was no difference. But not appropriate.

**Comment by Skrypt** [comment]
Made some changes hope you like it.

**Comment by Skrypt** [comment]
I need to find where Dapper binds the Table name schema or modify the BaseCommandInterpreter "I think". Else, it doesn't use the schema at all when creating tables. It interprets it as the table name simply.

**Comment by hishamco** [comment]
@Skrypt if you need any sort of help to test the PR in SQL Server or Postgres please ping me

**Comment by Skrypt** [comment]
image

**Comment by Skrypt** [comment]
Working now. Please test it. Happy new year 😉

**Comment by Skrypt** [comment]
Well, almost 😄

File: src/YesSql.Core/Sql/SchemaBuilder.cs [hunk 10]
```diff
@@ -36,7 +36,7 @@ private void Execute(IEnumerable<string> statements)
             foreach (var statement in statements)
             {
                 _logger.LogTrace(statement);
-                Connection.Execute(statement, null, Transaction);
+                var test = Connection.Execute(statement, null, Transaction);
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SchemaBuilder.cs:39]:
```suggestion
                Connection.Execute(statement, null, Transaction);
```

File: src/YesSql.Core/Sql/SchemaBuilder.cs [hunk 11]
```diff
@@ -205,10 +205,11 @@ public ISchemaBuilder AlterTable(string name, Action<IAlterTableCommand> table)
                 table(alterTable);
                 Execute(_commandInterpreter.CreateSql(alterTable));
             }
-            catch
+            catch(Exception ex)
             {
                 if (ThrowOnError)
                 {
+                    _logger.LogError(ex.Message);
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SchemaBuilder.cs:212]:
```suggestion
            catch
            {
                if (ThrowOnError)
                {
```

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 12]
```diff
@@ -79,7 +79,7 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
             // Don't prefix if alias is used
             if (alias != onTable)
             {
-                onTable = _tablePrefix + onTable;
+                onTable = _dialect.SchemaNameQuotedPrefix() + _tablePrefix + onTable;
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:82]: Can you check why there is no quoting on this table name?

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 13]
```diff
@@ -91,18 +91,20 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
...
+
             JoinSegments.Add(" INNER JOIN ");
-            JoinSegments.Add(_dialect.QuoteForTableName(_tablePrefix + table));
+            JoinSegments.Add(_dialect.SchemaNameQuotedPrefix() + _dialect.QuoteForTableName(_tablePrefix + table));
+
             if (!String.IsNullOrEmpty(alias))
             {
                 JoinSegments.AddRange(new[] { " AS ", _dialect.QuoteForTableName(alias) });
             }
+
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:102]: No schema here?

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 14]
```diff
@@ -158,9 +160,12 @@ public virtual string FormatColumn(string table, string column, bool isAlias = f
             if (!isAlias)
             {
                 table = _tablePrefix + table;
+                return _dialect.SchemaNameQuotedPrefix() + _dialect.QuoteForTableName(table) + "." + column;
+            }
+            else
+            {
+                return _dialect.QuoteForTableName(table) + "." + column;
```
Comment (sebastienros) [inline src/YesSql.Core/Sql/SqlBuilder.cs:167]: No schema here?

File: src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs [hunk 15]
```diff
@@ -83,8 +85,13 @@ static PostgreSqlDialect()
             };
         }
 
-        public PostgreSqlDialect()
+        public PostgreSqlDialect(string schema)
```
Comment (sebastienros) [inline src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs:88]: Should have a default `null` value

File: src/YesSql.Provider.SqlServer/SqlServerDialect.cs [hunk 16]
```diff
@@ -84,8 +84,13 @@ static SqlServerDialect()
             };
         }
 
-        public SqlServerDialect()
+        public SqlServerDialect(string schema)
```
Comment (sebastienros) [inline src/YesSql.Provider.SqlServer/SqlServerDialect.cs:87]: Default `null` value

**Review by sebastienros — COMMENTED** [review]
Very few locations where the schema is not prefixing table names, but I want to see what you have to say about my comments.

File: src/YesSql.Provider.SqlServer/SqlServerDialect.cs (hunk 16 above)
Comment (Skrypt) [reply src/YesSql.Provider.SqlServer/SqlServerDialect.cs:87]: default is "public"

File: src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs (hunk 15 above)
Comment (Skrypt) [reply src/YesSql.Provider.PostgreSql/PostgreSqlDialect.cs:88]: default is "public"

File: src/YesSql.Core/Sql/SqlBuilder.cs (hunk 13 above)
Comment (Skrypt) [reply src/YesSql.Core/Sql/SqlBuilder.cs:102]: Yes, there is probably an issue in here and on line 82. Will try to figure it out.

File: src/YesSql.Core/Sql/SqlBuilder.cs (hunk 14 above)
Comment (Skrypt) [reply src/YesSql.Core/Sql/SqlBuilder.cs:167]: It's an Alias so it doesn't take a Schema prefix. The real table names take a Schema name but once it is aliased it's not called with a schema.

**Comment by Skrypt** [comment]
Ok, now I'm getting to the strange part where it executes the tests fine one by one but not when I execute them all.

**Comment by Skrypt** [comment]
Maybe some tweaks here and there but mostly done and working 🚀

File: src/YesSql.Core/Provider/BaseDialect.cs [hunk 17]
```diff
@@ -193,6 +196,8 @@ public virtual string GetDropTableString(string name)
 
         public virtual bool PrefixIndex => false;
 
+        public virtual string Schema => String.Empty;
```
Comment (Skrypt) [inline src/YesSql.Core/Provider/BaseDialect.cs:199]: null or String.Empty?

File: src/YesSql.Provider.MySql/MySqlDialect.cs [hunk 18]
```diff
@@ -248,6 +248,9 @@ public override string QuoteForTableName(string tableName)
             return "`" + tableName + "`";
         }
 
+        // MySQL doesn't have this feature
+        public override string SchemaNameQuotedPrefix() => String.Empty;
```
Comment (Skrypt) [inline src/YesSql.Provider.MySql/MySqlDialect.cs:252]: Null or String.Empty?

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 19]
```diff
@@ -79,30 +79,40 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
             // Don't prefix if alias is used
```
Comment (Skrypt) [inline src/YesSql.Core/Sql/SqlBuilder.cs:79]: As noted here I removed prefixes for alias so either I revert this change or I keep it and we create a migration in Orchard Core.

File: src/YesSql.Core/Sql/SqlBuilder.cs [hunk 20]
```diff
@@ -79,30 +79,40 @@ public virtual void InnerJoin(string table, string onTable, string onColumn, str
...
+            }
+            else
             {
                 toTable = _tablePrefix + toTable;
             }
 
             if (!String.IsNullOrEmpty(toAlias))
             {
-                toTable = toAlias;
+                toTable = _dialect.QuoteForTableName(toAlias);
```
Comment (Skrypt) [inline src/YesSql.Core/Sql/SqlBuilder.cs:100]: Here I needed to add the `QuoteForTableName(toAlias)` because it was previously done all the time on line 113. So Alias were using `QuoteForTableName()`. Also, it is used on Alias because we also do it on line 108 but I'm not sure it is necessary to add Quotes on an Alias name in an INNER JOIN query; We could also try to remove QuoteForTableName() on lines 100 and 108 to see if tests pass.

Comment (Skrypt) [reply src/YesSql.Core/Sql/SqlBuilder.cs]: Seems like it's working perfectly fine without quotes.

Comment (Skrypt) [reply src/YesSql.Core/Sql/SqlBuilder.cs]: Well, I talked too quickly. PostgreSQL doesn't like it.

**Comment by Skrypt** [comment]
Here is the issue with PostgreSQL : 

```SQL
SELECT BabyYoda."tpDocument".* FROM BabyYoda."tpDocument" INNER JOIN BabyYoda."tpArticlesByDay_Document" AS ArticlesByDay_Document_a1 ON ArticlesByDay_Document_a1."DocumentId" = BabyYoda."tpDocument"."Id" INNER JOIN BabyYoda."tpArticlesByDay" AS ArticlesByDay_a1 ON ArticlesByDay_a1."Id" = ArticlesByDay_Document_a1."ArticlesByDayId" WHERE ("ArticlesByDay_a1"."DayOfYear" = @p0) limit 1 limit 1 limit 1
```

The `WHERE` clause is not using the same `ALIAS` name as to how it is declared in the `INNER JOIN`. So the issue with PostgreSQL is that you need to be consistent by using the same ALIAS name everywhere in the Query. So, it supports alias quotes but it is not necessary. The QuoteForAliasName here is kind of a patch because it forces using quotes on an ALIAS name whereas the WHERE clause should be smart enough to know that the table it refers to is actually an ALIAS and thus not add quotes if it was not added previously when doing AS.

**Comment by lnaie** [comment]
This is something I noticed was missing. Is it gonna make it?

**Comment by lnaie** [comment]
Come on, guys! :) The PRs are pilling up.

**Comment by Skrypt** [comment]
You can use the preview feed if you need it now.

**Comment by Skrypt** [comment]
Ok just analyzed this and the MyGet feed only has what is pushed on the main branch. PR's are not pushed on MyGet which makes sense. So you will need to get it from my fork and create the Nuget packages and use them by adding the local folder in your nuget.config file if you want to use it now.

**Comment by Skrypt** [comment]
By the way @sebastienros the PR is ready. Even though some things that I noted could be refactored this PR is ready.

**Comment by Skrypt** [comment]
This PR is now 9 months old. I stopped progress on this mainly because we never had time to review/merge this.
@sebastienros

**Comment by hishamco** [comment]
@Skrypt if there's anything with this PR I can do, let me know

**Comment by Skrypt** [comment]
Nothing much. We just need to have time.

**Comment by sebastienros** [comment]
Refactored it so that dialects don't depend on the schema (will be useful for OC too) but is part of the configuration.

Tests are failing here with PG but worked on my machine, maybe an issue if the schema doesn't exist yet. Related to that I think the InitializeAsync method should create the schema too if it doesn't exist. Might be helpful. Users can still create it if they want beforehand (with custom arguments).

**Comment by sebastienros** [comment]
Could be valuable to add a unit test for PG that doesn't set the Schema, to be sure that it works without (not even setting the default value 'public'). Done by creating a subclass of PgSqlTest.

**Comment by Skrypt** [comment]
Are you sure that MySQL supports Schemas as PG and SQL Server do?

**Comment by Skrypt** [comment]
Good work @sebastienros 😄
